File: packages/core/src/tracing.ts
import { Logger } from "iii-sdk";
import type { Memory } from "./memory.js";

export interface SpanEvent {
  agentId: string;
  runId: string;
  parentRunId?: string;
  event: string;
  data?: unknown;
  ts: number;
}

export const TRACE_STREAM = "snoopy.trace";

/**
 * Pluggable trace sink. Implementations: stdout (always on), iii OTel
 * (always on, falls back silently), Memory stream (when memory passed),
 * and any user-provided sinks (PgTraceStore, custom adapters).
 */
export interface TraceSink {
  record(span: SpanEvent): void | Promise<void>;
}

const externalSinks: TraceSink[] = [];

/**
 * Register an extra trace sink (e.g. `addTraceSink(new PgTraceStore())` for
 * durable spans). Sinks are best-effort: failures are swallowed.
 */
export function addTraceSink(sink: TraceSink): void {
  externalSinks.push(sink);
}

let _logger: Logger | undefined;
function logger(): Logger {
  return (_logger ??= new Logger(undefined, "snoopy"));
}

export class Span {
  constructor(
    public readonly agentId: string,
    public readonly runId: string,
    public readonly parentRunId?: string,
  ) {}
}

/**
 * Emit a span. Goes to four sinks in best-effort order:
 *   1. stdout (so `agent dev` shows spans inline)
 *   2. iii OTel logger (distributed tracing dashboards pick it up)
 *   3. memory append (if `memory` provided — Redis stream for `agent traces`)
 *   4. external sinks registered via addTraceSink (Postgres, custom)
 *
 * Tracing must never break the agent loop. Each sink is guarded.
 */
export function emitSpan(
  span: Span,
  payload: { event: string; data?: unknown },
  memory?: Memory,
): void {
  const ev: SpanEvent = {
    agentId: span.agentId,
    runId: span.runId,
    parentRunId: span.parentRunId,
    event: payload.event,
    data: payload.data,
    ts: Date.now(),
  };

  if (process.env.SNOOPY_TRACE_STDOUT !== "false") {
    const tag = `[trace ${ev.agentId} ${ev.runId.slice(0, 8)}]`;
    // eslint-disable-next-line no-console
    console.log(tag, ev.event, ev.data ?? "");
  }

  try {
    logger().info(`snoopy.span ${ev.event}`, ev as unknown as Record<string, unknown>);
  } catch {}

  if (memory) {
    void memory.append(TRACE_STREAM, ev).catch(() => {});
  }

  for (const sink of externalSinks) {
    Promise.resolve(sink.record(ev)).catch(() => {});
  }
}

File: python/snoopy/tracing.py
"""Span emission — stdout + pluggable external sinks."""
from __future__ import annotations

import os
import time
from dataclasses import dataclass, field
from typing import Any, Optional, Protocol

TRACE_STREAM = "snoopy.trace"


@dataclass
class Span:
    agent_id: str
    run_id: str
    parent_run_id: Optional[str] = None


@dataclass
class SpanEvent:
    agent_id: str
    run_id: str
    event: str
    ts: int = field(default_factory=lambda: int(time.time() * 1000))
    data: Any = None
    parent_run_id: Optional[str] = None

    def to_dict(self) -> dict[str, Any]:
        return {
            "agentId": self.agent_id,
            "runId": self.run_id,
            "parentRunId": self.parent_run_id,
            "event": self.event,
            "data": self.data,
            "ts": self.ts,
        }


class TraceSink(Protocol):
    async def record(self, event: SpanEvent) -> None: ...


_external_sinks: list[TraceSink] = []


def add_trace_sink(sink: TraceSink) -> None:
    _external_sinks.append(sink)


def emit_span(span: Span, *, event: str, data: Any = None, memory=None) -> None:
    """
    Best-effort span emission. Goes to stdout, optional Memory stream
    (Redis), and any registered external sinks. Never raises.
    """
    ev = SpanEvent(
        agent_id=span.agent_id,
        run_id=span.run_id,
        parent_run_id=span.parent_run_id,
        event=event,
        data=data,
    )

    if os.environ.get("SNOOPY_TRACE_STDOUT", "true").lower() not in ("false", "0"):
        run_short = ev.run_id[:8] if ev.run_id else ""
        print(f"[trace {ev.agent_id} {run_short}] {ev.event} {data or ''}")

    if memory is not None:
        try:
            import asyncio

            asyncio.create_task(_safe_append(memory, ev))
        except Exception:
            pass

    for sink in _external_sinks:
        try:
            import asyncio

            asyncio.create_task(_safe_record(sink, ev))
        except Exception:
            pass


async def _safe_append(memory, ev: SpanEvent) -> None:
    try:
        await memory.append(TRACE_STREAM, ev.to_dict())
    except Exception:
        pass


async def _safe_record(sink: TraceSink, ev: SpanEvent) -> None:
    try:
        await sink.record(ev)
    except Exception:
        pass
